#414: slow StyledTextRenderer for unicode and unprintable control cha…

tmssngr opened the pull request:
…racters

This test replaces the characters 0x00-0x1F with some alternative character (by default '?') for rendering. This improves the performance significantly.

The previous pull request (https://github.com/eclipse-platform/eclipse.platform.swt/pull/1425) did not work.

Note, that when changing `replaceBelowSpaceCharacters` to return the commented unicode characters, the snippet will fail with a "No more handles" exception. I reckon, we best should stick to a US-ASCII character from the range 0x20<=..<0x7F.

merks commented on bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/TextLayout.java:
```diff
@@ -2854,6 +2854,13 @@ public boolean isDisposed () {
+	return chr >= ' '
+		? chr
+		//: (char)(0x2400 + chr);
+		: '?';
```
Given that \t\r\n all fall into this category, can this really be correct?

tmssngr replied:
You are right. The bug was not directly visible, but caused weird effects when editing.

tmssngr commented:
My commit used the correct, approved email address, but I've created the pull request with my GitHub account. Don't know why eclipsefdn/eca complains.

iloveeclipse commented:
> My commit used the correct, approved email address, but I've created the pull request with my GitHub account. Don't know why eclipsefdn/eca complains.

This commit should only have committers / authors with ECA signed, but you see there two different identities : https://github.com/syntevo/eclipse.platform.swt/commit/176f48ac36cfb00fa5db757f0b6e3fb5b88f885d

ECA wants clarity regarding the code origin, so that **every** contribution is legal.

tmssngr committed:
#414: slow StyledTextRenderer for unicode and unprintable control characters

This test replaces the characters 0x00-0x1F with some alternative
character (by default '?') for rendering. This improves the performance
significantly.

mihnita commented:
I wonder if it has something to do with the "Interpret ASCII control characters" setting
(under Settings -- Run/Debug -- Console)

I know that is only about the Console, but the console is implemented on top of StyledText.
So maybe the root cause is that processing.

mihnita commented on bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/TextLayout.java:
```diff
@@ -2854,6 +2854,14 @@ public boolean isDisposed () {
 }
 
+private static char replaceBelowSpaceCharacters(char chr) {
+	if (chr >= 0x20 || chr == '\t' || chr == '\r' || chr == '\n') {
```
Should we consider allowing more?
Other characters that are used in CLI applications:
* `\u0007` : BELL, makes a "beep" sound
* `\u0008` : BS, backspace, moves back one character
* `\u001B` : ESC, used for ANSI escape sequences (changing colors, cursor movement, clear screen, etc). This one is used quite a lot.

Since the Eclipse Console is built on top of the styled text widget not allowing ESC might break quite a lot of things.
(see https://eclipse.dev/eclipse/news/4.25/platform.php#debug-ansi-support)
